frameworks/shared/serialization.py

```diff
@@ -28,16 +32,71 @@ def _import_data_libraries():
     pandas_compression=os.environ.get('AMLB_SER_PD_COMPR') or 'infer',
     # 'snappy', 'gzip', 'brotli', None ?
     pandas_parquet_compression=os.environ.get('AMLB_SER_PD_PQT_COMPR') or None,
+    # if numpy can use pickle to serialize ndarrays,
+    numpy_pickle=True,
+    # if sparse matrices should be compressed during serialization.
+    sparse_matrix_compression=True,
+    # if sparse matrices should be deserialized to some specific format:
+    #  None (no change), 'array' (numpy), 'dense' (dense matrix).
+    sparse_matrix_deserialized_format='array',
+    # if sparse dataframes should be deserialized to some specific format:
+    #  None (no change), 'array' (numpy), 'dense' (dense dataframe/series).
+    sparse_dataframe_deserialized_format=None
 )
 
 
 __series__ = '_series_'
 
 
+def is_serializable_data(data):
+    np, pd, sp = _import_data_libraries()
+    return isinstance(data, (np.ndarray, sp.spmatrix, pd.DataFrame, pd.Series))
+
+
+def unsparsify(*data, fmt=None):
+    if len(data) == 1:
+        return _unsparsify(data[0], fmt=fmt)
+    else:
+        return tuple(_unsparsify(d, fmt=fmt) for d in data)
+
+
+def _unsparsify(data, fmt=None):
+    """
+    :param data: the matrix to process.
+    :param fmt: one of None, 'array', 'dense'
+    :return: the original matrix is fmt is None,
+            a numpy array if fmt is 'array',
+            a dense version of the data type if fmt is 'dense'.
+    """
+    if fmt is None:
+        return data
+    np, pd, sp = _import_data_libraries()
+    if sp and isinstance(data, sp.spmatrix):
+        return (data.toarray() if fmt == 'array'
+                else data.todense() if fmt == 'dense'
+                else data)
+    elif pd and isinstance(data, (pd.DataFrame, pd.Series)):
+        return (data.to_numpy() if fmt == 'array'
+                else data.sparse.to_dense() if fmt == 'dense' and hasattr(data, 'sparse')
+                else data)
+    else:
+        return data
+
+
 @profile(log)
 def serialize_data(data, path):
-    np, pd = _import_data_libraries()
-    if pd and isinstance(data, (pd.DataFrame, pd.Series)):
+    np, pd, sp = _import_data_libraries()
+    if np and isinstance(data, np.ndarray):
```

Comment:
The added check on arrays makes MLPlan fails which procudes `list` predictions (and probabilities).

Reply:
serialization module was never supposed to serialize lists… that's why the methods are named `serialized_data`, accepting only data formats.
Lists go with the blob in json format… I'll probably just fix MLPlan to convert predictions to array.

Reply:
Since `result` does not have a signature I wasn't sure whether list format predictions/probabilities were officially supported but assumed they were. If the support was only accidental then changing MLPlan integration is the correct call 👍

Reply:
resolved: serialization now also supports basic datastructures